### DISCUSSION TITLE
Adding MPI Support

### DISCUSSION THREAD

**Discussion by jtronge** [lanl/BEE#377] [Ideas]
I'm creating this discussion to talk about how we can add some level of MPI support to BEE. I'm thinking this would require some modifications to the Worker code in order to wrap steps with the proper `mpirun` command. It may also require some modifications to the container runtime classes, but we might be able to get around that. 

Also, we'll probably need to have some sort of new hint or requirement for MPI. The CWL team published a [paper](https://arxiv.org/pdf/2010.00422.pdf) last year about a `cwltool:MPIRequirement` extension that they experimented with. Perhaps we can use this requirement to some extent, or come up with something else that's better suited to HPC.

**Comment by pagrubel** [comment]
For updating for Slurm and Charliecloud it looks fairly simple see:
https://hpc.github.io/charliecloud/tutorial.html#non-interactive-mpi-hello-world

**Comment by jtronge** [comment]
Using Singularity+MPI looks to be fairly straightforward based on their docs: https://sylabs.io/guides/3.3/user-guide/mpi.html. It does look like the container build process, such as for including the MPI libraries, may be more complicated.

**Comment by jtronge** [comment]
I wanted to make a note about Quincy's comment for using prefixes for requirements/hints in order to avoid polluting the CWL namespace. I found an example of this with [Arvados](https://doc.arvados.org/user/cwl/cwl-extensions.html). They use an `arv:` prefix. I think we might be able to do something like `beeflow:SomeRequirement` like they are doing.

**Comment by jtronge** [comment]
After looking over the worker and container runtime code some more, I think we may need to modify the container runtime interface. Right now the container code simply generates script text to be appended onto the job submission scripts. This makes prepending to the main command difficult within the worker. An example of why this seems to be necessary is in the [Charliecloud link](https://hpc.github.io/charliecloud/tutorial.html#non-interactive-mpi-hello-world) that Pat posted above. You can see that there are some commands that just set up the environment, like the `module load` commands and then other commands that need to have `srun` prepended in order to run on the allocated nodes, if I understand the example correctly.

I'm thinking that maybe instead of generating text, the container runtime could instead generate a list of special `Command` objects with specific properties (in a new method `commands(task)`). Then the worker would use each special command to build the final submission script. Does that seem like the best way to do this?

**Reply by rstyd** [reply]
Hey Jake,
I'll have to check out the code a bit more, but that sounds like a solid plan to me. The granularity is definitely too low at the moment.

**Comment by pagrubel** [comment]
This does sound like a good idea. I'm not sure what specific properties would be, something like pre-pend?

**Reply by jtronge** [reply]
I'm thinking that `Command` object might have some sort of `type` property. We might have an `environment` type that corresponds to environment commands like `module load charliecloud` and then a `compute` or `all-node` type that would correspond the `ch-tar2dir ...` and `ch-run ...` commands that should be prepended.

**Reply by pagrubel** [reply]
We'd also need an order 
for instance on the main compute node you may run
srun ch-tar2dir
then
srun --cpus-per-task=1 ch-run ...

**Reply by jtronge** [reply]
Yeah. The container runtime will probably need to return an in-order list of `Command` objects. These will then be concatenated in order by the worker.

**Comment by kinow** [comment]
Hi, I hope it's ok that I'm commenting on this old discussion. I am working with @mr-c on my master thesis about CWL and MPI, checking the current status of `MPIRequirement` in cwltool, discussing adding it to the CWL specification, and comparing how CWL implementations are using MPI.

@mr-c told me about BEE and that it support CWL and MPI, so over next days I will perform some quick tests with BEE + MPI, but if there is anyone still working on this discussion/issue, with ideas or requirements, I'd be keen to discuss possible ways to improve the CWL implementation, or maybe include it in my master thesis. I intend to test some CWL workflows on different HPC and cloud providers, with cwltool, toil, and other CWL implementations.

Cheers

**Comment by pagrubel** [comment]
@kinow I would suggest you look at the pre-release version of BEE. You can install it using pip. The documentation is at this site https://lanl.github.io/BEE/0.1.11dev5/ 
BEE requires either a spack package (new feature) with redis installed in the spack environment, and if you select using SQLite for the graph database (default) containers are not required.
Or the more typical installation uses Charliecloud and containers for redis and neo4j.
If you have any questions, feel free to email bee-dev@lanl.gov

**Reply by kinow** [reply]
Thank you, @pagrubel !
